scylla/src/transport/load_balancing/default.rs

```diff
@@ -1563,4 +1583,315 @@ mod tests {
+                    .build(),
+            },
+            Test {
+                // No keyspace implies no token awareness
+                policy: DefaultPolicy {
+                    preferred_datacenter: Some("eu".to_owned()),
+                    is_token_aware: true,
+                    permit_dc_failover: true,
+                    ..Default::default()
+                },
+                query_info: QueryInfo {
+                    token: Some(Token { value: 160 }),
+                    keyspace: None, // no keyspace
+                    consistency: Consistency::Quorum,
+                    ..Default::default()
+                },
+                expected_groups: ExpectedGroupsBuilder::new()
+                    .group([A, B, C, G]) // local nodes
+                    .group([D, E, F]) // remote nodes
+                    .build(),
+            },
+            Test {
+                // Unknown preferred DC, failover permitted
+                policy: DefaultPolicy {
+                    preferred_datacenter: Some("au".to_owned()),
+                    is_token_aware: true,
+                    permit_dc_failover: true,
+                    ..Default::default()
+                },
+                query_info: QueryInfo {
+                    token: Some(Token { value: 160 }),
+                    keyspace: Some(KEYSPACE_NTS_RF_2),
+                    consistency: Consistency::Quorum,
+                    ..Default::default()
+                },
+                // going though the ring, we get order: F , A , C , D , G , B , E
+                //                                      us  eu  eu  us  eu  eu  us
+                //                                      r2  r1  r1  r1  r2  r1  r1
+                expected_groups: ExpectedGroupsBuilder::new()
+                    .group([A, D, F, G]) // remote replicas
+                    .group([B, C, E]) // remote nodes
+                    .build(),
+            },
+            Test {
+                // Unknown preferred DC, failover forbidden
+                policy: DefaultPolicy {
+                    preferred_datacenter: Some("au".to_owned()),
+                    is_token_aware: true,
+                    permit_dc_failover: false,
+                    ..Default::default()
+                },
+                query_info: QueryInfo {
+                    token: Some(Token { value: 160 }),
+                    keyspace: Some(KEYSPACE_NTS_RF_2),
+                    consistency: Consistency::Quorum,
+                    ..Default::default()
+                },
+                // going though the ring, we get order: F , A , C , D , G , B , E
+                //                                      us  eu  eu  us  eu  eu  us
+                //                                      r2  r1  r1  r1  r2  r1  r1
+                expected_groups: ExpectedGroupsBuilder::new().build(), // empty plan, because all nodes are remote and failover is forbidden
+            },
```

Comment:
Maybe it could also be tested what happens when the keyspace is not available.

Reply:
It is already tested in one test, in line 1815 (look for `// No keyspace implies no token awareness`)

Reply:
Right, missed it :+1: